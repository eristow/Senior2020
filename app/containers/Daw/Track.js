import React, { useEffect } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import TrackConfig from './TrackConfig';
import Steps from './Steps';

const Container = styled.div`
  display: flex;
  background: gray;
  width: auto;
  height: auto;
  border: 1px solid black;
  margin: 1px;
  position: relative;
`;

const IndicatorWrapper = styled.div`
  position: absolute;
  width: 570px;
  top: 0px;
  right: 0px;
  left: 100px;
  height: 100px;
  padding: 4px 2px;
  pointer-events: none;
`;

// TODO: how to make this look smoother? interpolate between last location and current one?
const StepIndicator = styled.div`
  position: absolute;
  top: 0px;
  left: ${props => props.step * 41.5}px;
  width: 40px;
  height: 100%;
  background: #00ff0040;
`;

export function Track({
  name,
  num,
  playing,
  currentStep,
  stepState,
  buffer,
  setBuffers,
}) {
  useEffect(() => {
    setBuffers(buffers => ({
      ...buffers,
      [name]: buffer,
    }));
  }, [buffer]);

  return (
    <Container>
      <TrackConfig num={num} buffer={buffer} />
      {/* TODO: this is where patterns will go (midi or audio) */}
      <IndicatorWrapper>
        {playing && <StepIndicator step={currentStep} />}
      </IndicatorWrapper>
      <Steps name={name} stepsState={stepState} />
    </Container>
  );
}

Track.propTypes = {
  name: PropTypes.string,
  num: PropTypes.number,
  playing: PropTypes.bool,
  currentStep: PropTypes.number,
  stepState: PropTypes.object,
  buffer: PropTypes.object,
  setBuffers: PropTypes.func,
};

export default Track;
