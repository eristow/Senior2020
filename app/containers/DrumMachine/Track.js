import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';
import { useInjectReducer } from 'utils/injectReducer';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import Slider from 'components/Slider';
import { makeSelectTrackVol } from './selectors';
import { changeTrackVol } from './actions';
import reducer from './reducer';

import Steps from './Steps';

const Wrapper = styled.div`
  display: flex;
  flex: 1;
`;

const Info = styled.div`
  flex: 0 0 155px;
  background: linear-gradient(#555, #666);
  border: 1px solid #555;
`;

const Name = styled.h2`
  color: white;
  font-size: 14px;
  margin: 0;
  vertical-align: middle;
  padding: 0px 10px;
  line-height: 50px;
`;

const key = 'drumMachine';

export function Track({
  onChangeTrackVol,
  buffer,
  name,
  setBuffers,
  trackVol,
}) {
  useInjectReducer({ key, reducer });

  useEffect(() => {
    // I don't know if there's a better way to do this to make eslint happy
    // eslint-disable-next-line no-param-reassign
    buffer.volume.value = trackVol[name];
  }, [trackVol]);

  useEffect(() => {
    setBuffers(buffers => ({
      ...buffers,
      [name]: buffer,
    }));
  }, [buffer]);

  return (
    <Wrapper>
      <Info>
        <Name>{name}</Name>
        <Slider
          onChange={e => onChangeTrackVol(name, e)}
          min={-55}
          max={0}
          value={trackVol[name]}
          width={120}
        />
      </Info>
      <Steps name={name} />
    </Wrapper>
  );
}

Track.propTypes = {
  onChangeTrackVol: PropTypes.func,
  buffer: PropTypes.object,
  name: PropTypes.string,
  setBuffers: PropTypes.func,
  trackVol: PropTypes.object,
};

const mapStateToProps = createStructuredSelector({
  trackVol: makeSelectTrackVol(),
});

const mapDispatchToProps = dispatch => ({
  onChangeTrackVol: (name, evt) => {
    dispatch(changeTrackVol(name, evt));
  },
});

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export default compose(withConnect)(Track);
