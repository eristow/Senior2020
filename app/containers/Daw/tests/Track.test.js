import React from 'react';
import { render } from 'react-testing-library';
import { Provider } from 'react-redux';
import configureStore from '../../../configureStore';

import { Track } from '../Track';

describe('<Track />', () => {
  let store;
  const stepState = {
    Track1: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    Track2: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    Track3: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
    Track4: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
  };
  const name = 'Track1';
  const buffer = { volume: { value: 0 } };
  const setBuffers = jest.fn();

  beforeAll(() => {
    store = configureStore();
  });

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    const dispatch = jest.fn();

    render(
      <Provider store={store}>
        <Track
          dispatch={dispatch}
          name={name}
          stepState={stepState}
          buffer={buffer}
          setBuffers={setBuffers}
        />
      </Provider>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <Track
          name={name}
          stepState={stepState}
          buffer={buffer}
          setBuffers={setBuffers}
        />
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });
});
