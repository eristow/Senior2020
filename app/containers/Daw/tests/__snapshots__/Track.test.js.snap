// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Track /> Should render and match the snapshot 1`] = `
<div
  class="Track__Container-dcw2r7-0 etskRi"
>
  <div
    class="TrackConfig__Container-ouodsg-0 ddcwJq"
  >
    <input
      class="InputText-wtwjzw-0 ihxodh"
      font-size="0.75em"
      type="text"
      value="Track1"
      width="97%"
    />
    <button
      class="TrackConfig__Select-ouodsg-1 cstDwF"
      type="button"
    >
      Select
    </button>
    <div>
      <div
        style="display: inline-block;"
      >
        <div
          class="rc-slider"
          style="width: 75px; height: 5px; margin: 0.5em 0.5em; z-index: 0;"
        >
          <div
            class="rc-slider-rail"
          />
          <div
            class="rc-slider-track"
            style="left: 0%; width: 100%;"
          />
          <div
            class="rc-slider-step"
          />
          <div
            aria-disabled="false"
            aria-valuemax="0"
            aria-valuemin="-60"
            aria-valuenow="0"
            class="rc-slider-handle"
            role="slider"
            style="left: 100%; transform: translateX(-50%);"
            tabindex="0"
          />
          <div
            class="rc-slider-mark"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="Track__IndicatorWrapper-dcw2r7-1 cztnBH"
  />
  <div
    class="Steps__Wrapper-sc-8e49i7-0 iDvuHY"
  >
    <button
      class="Step__StepButton-sc-1gy89br-0 buOaSX"
    />
    <button
      class="Step__StepButton-sc-1gy89br-0 buOaSX"
    />
    <button
      class="Step__StepButton-sc-1gy89br-0 buOaSX"
    />
    <button
      class="Step__StepButton-sc-1gy89br-0 buOaSX"
    />
    <button
      class="Step__StepButton-sc-1gy89br-0 iNYnBq"
    />
    <button
      class="Step__StepButton-sc-1gy89br-0 iNYnBq"
    />
    <button
      class="Step__StepButton-sc-1gy89br-0 iNYnBq"
    />
    <button
      class="Step__StepButton-sc-1gy89br-0 iNYnBq"
    />
    <button
      class="Step__StepButton-sc-1gy89br-0 buOaSX"
    />
    <button
      class="Step__StepButton-sc-1gy89br-0 buOaSX"
    />
    <button
      class="Step__StepButton-sc-1gy89br-0 buOaSX"
    />
    <button
      class="Step__StepButton-sc-1gy89br-0 buOaSX"
    />
    <button
      class="Step__StepButton-sc-1gy89br-0 iNYnBq"
    />
    <button
      class="Step__StepButton-sc-1gy89br-0 iNYnBq"
    />
    <button
      class="Step__StepButton-sc-1gy89br-0 iNYnBq"
    />
    <button
      class="Step__StepButton-sc-1gy89br-0 iNYnBq"
    />
  </div>
</div>
`;
